core/primitives/src/transaction.rs

```diff
@@ -49,6 +70,147 @@ impl Transaction {
+        match self {
+            Transaction::V0(tx) => &tx.signer_id,
+            Transaction::V1(tx) => &tx.signer_id,
+        }
+    }
+
+    pub fn receiver_id(&self) -> &AccountId {
+        match self {
+            Transaction::V0(tx) => &tx.receiver_id,
+            Transaction::V1(tx) => &tx.receiver_id,
+        }
+    }
+
+    pub fn public_key(&self) -> &PublicKey {
+        match self {
+            Transaction::V0(tx) => &tx.public_key,
+            Transaction::V1(tx) => &tx.public_key,
+        }
+    }
+
+    pub fn nonce(&self) -> Nonce {
+        match self {
+            Transaction::V0(tx) => tx.nonce,
+            Transaction::V1(tx) => tx.nonce,
+        }
+    }
+
+    pub fn actions(&self) -> &[Action] {
+        match self {
+            Transaction::V0(tx) => &tx.actions,
+            Transaction::V1(tx) => &tx.actions,
+        }
+    }
+
+    pub fn take_actions(self) -> Vec<Action> {
+        match self {
+            Transaction::V0(tx) => tx.actions,
+            Transaction::V1(tx) => tx.actions,
+        }
+    }
+
+    pub fn block_hash(&self) -> &CryptoHash {
+        match self {
+            Transaction::V0(tx) => &tx.block_hash,
+            Transaction::V1(tx) => &tx.block_hash,
+        }
+    }
+
+    pub fn priority_fee(&self) -> Option<u64> {
+        match self {
+            Transaction::V0(_) => None,
+            Transaction::V1(tx) => Some(tx.priority_fee),
+        }
+    }
+}
+
+impl BorshSerialize for Transaction {
+    fn serialize<W: Write>(&self, writer: &mut W) -> Result<(), Error> {
+        match self {
+            Transaction::V0(tx) => tx.serialize(writer)?,
+            Transaction::V1(tx) => {
+                BorshSerialize::serialize(&1_u8, writer)?;
+                tx.serialize(writer)?;
+            }
+        }
+        Ok(())
+    }
+}
+
+impl BorshDeserialize for Transaction {
+    /// Deserialize based on the first and second bytes of the stream. For V0, we do backward compatible deserialization by deserializing
+    /// the entire stream into V0. For V1, we consume the first byte and then deserialize the rest.
+    fn deserialize_reader<R: Read>(reader: &mut R) -> std::io::Result<Self> {
+        let u1 = u8::deserialize_reader(reader)?;
+        let u2 = u8::deserialize_reader(reader)?;
+        let u3 = u8::deserialize_reader(reader)?;
+        let u4 = u8::deserialize_reader(reader)?;
+        // This is a ridiculous hackery: because the first field in `TransactionV0` is an `AccountId`
+        // and an account id is at most 64 bytes, for all valid `TransactionV0` the second byte must be 0
+        // because of the littel endian encoding of the length of the account id.
+        // On the other hand, for `TransactionV1`, since the first byte is 1 and an account id must have nonzero
+        // length, so the second byte must not be zero. Therefore, we can distinguish between the two versions
+        // by looking at the second byte.
```

Comment:
> Is my understanding correct that using 1 as the sentinel value also relies on the range of valid account lengths to be between 2 and 64, rather than just <= 64? If so, it would make sense to add that to the comment.


No 1 is not a sentinel value. Rather, it is because the `V1` enum serialization writes 1 first. You can think of `0` as the sentinel value in the opposite sense - this byte is 0 if and only if the transaction is v0. For subsequent versions, this byte denotes which version it is

Reply:
but wouldn't it already break down for version 2, as it collides with version 0 and account ids of length 2?

Reply:
Yes I didn't find a clean way to do it. That is why I call this piece of code a ridiculous hackery 😅